kasa/discover.py

```diff
@@ -193,22 +224,52 @@ async def discover(
         return protocol.discovered_devices
 
     @staticmethod
-    async def discover_single(host: str, *, port: Optional[int] = None) -> SmartDevice:
+    async def discover_single(
+        host: str, *, port: Optional[int] = None, timeout=5
+    ) -> SmartDevice:
         """Discover a single device by the given IP address.
 
         :param host: Hostname of device to query
         :rtype: SmartDevice
         :return: Object for querying/controlling found device.
         """
-        protocol = TPLinkSmartHomeProtocol(host, port=port)
-
-        info = await protocol.query(Discover.DISCOVERY_QUERY)
-
-        device_class = Discover._get_device_class(info)
-        dev = device_class(host, port=port)
-        await dev.update()
+        loop = asyncio.get_event_loop()
+        transport, protocol = await loop.create_datagram_endpoint(
+            lambda: _DiscoverProtocol(target=host, port=port),
+            local_addr=("0.0.0.0", 0),
+        )
+        protocol = cast(_DiscoverProtocol, protocol)
+        timer = 0.0
+        wait = 0.0001
+        try:
+            _LOGGER.debug("Waiting a total of %s seconds for responses...", timeout)
+            while (
+                timer < timeout
+                and host not in protocol.discovered_devices
+                and host not in protocol.unsupported_devices
+                and host not in protocol.invalid_device_exceptions
+            ):
+                _LOGGER.debug("Sleeping for %s seconds", wait)
+                await asyncio.sleep(wait)
+                timer += wait
+                wait = min(wait * 10, 1)
```

Comment:
It would be better to use `async_timeout` (see protocol.py) and `asyncio.Event` inside the `_DiscoverProtocol` to inform that the discovery is done.

It's not a must, and we can refine it later, but pleases remove the debug logging at least as that is otherwise very spammy.

Reply:
Done as per suggestion with `async_timeout` and `asyncio.Event`